packages/thirdweb/src/wallets/in-app/core/authentication/passkeys.ts

```diff
@@ -0,0 +1,199 @@
+    rp: RpInfo;
+  }) => Promise<AuthenticateResult>;
+  isAvailable: () => boolean;
+}
+
+export async function registerPasskey(options: {
+  client: ThirdwebClient;
+  storage: AsyncStorage;
+  passkeyClient: PasskeyClient;
+  ecosystem?: Ecosystem;
+  username?: string;
+  rp: RpInfo;
+}): Promise<AuthStoredTokenWithCookieReturnType> {
+  if (!options.passkeyClient.isAvailable()) {
+    throw new Error("Passkeys are not available on this device");
+  }
+  const storage = new LocalStorage({
+    storage: options.storage,
+    clientId: options.client.clientId,
+    ecosystemId: options.ecosystem?.id,
+  });
+  const fetchWithId = getClientFetch(options.client, options.ecosystem);
+  const generatedName = options.username ?? generateUsername(options.ecosystem);
+  // 1. request challenge from  server
+  const res = await fetchWithId(getChallengePath("sign-up", generatedName));
+  const challengeData = await res.json();
+  if (!challengeData.challenge) {
+    throw new Error("No challenge received");
+  }
+  const challenge = challengeData.challenge;
+
+  console.log("DEBUG -- challengeData", challenge);
+  // 2. initiate registration
+  const registration = await options.passkeyClient.register({
+    name: generatedName,
+    challenge,
+    rp: options.rp,
+  });
+  console.log("DEBUG -- registration", registration);
+  // 3. store the credentialId in local storage
+  await storage.savePasskeyCredentialId(registration.credentialId);
+
+  const customHeaders: Record<string, string> = {};
+  if (options.ecosystem?.partnerId) {
+    customHeaders["x-ecosystem-partner-id"] = options.ecosystem.partnerId;
+  }
+  if (options.ecosystem?.id) {
+    customHeaders["x-ecosystem-id"] = options.ecosystem.id;
+  }
+
+  // 4. send the registration object to the server
+  const verifRes = await fetchWithId(getVerificationPath(), {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+      ...customHeaders,
+    },
+    body: JSON.stringify({
+      type: "sign-up",
+      authenticatorData: registration.authenticatorData,
+      credentialId: registration.credentialId,
+      serverVerificationId: challengeData.serverVerificationId,
+      clientData: registration.clientData,
+      username: generatedName,
+      credential: {
+        publicKey: registration.credential.publicKey,
+        algorithm: registration.credential.algorithm,
+      },
+    }),
+  });
+  const verifData = await verifRes.json();
+
+  if (!verifData) {
+    throw new Error("No token received");
+  }
+  // 5. returns back the IAW authentication token
+  return verifData;
+}
+
+export async function loginWithPasskey(options: {
+  client: ThirdwebClient;
+  storage: AsyncStorage;
+  passkeyClient: PasskeyClient;
+  rp: RpInfo;
+  ecosystem?: Ecosystem;
+}): Promise<AuthStoredTokenWithCookieReturnType> {
+  if (!options.passkeyClient.isAvailable()) {
+    throw new Error("Passkeys are not available on this device");
+  }
+  const storage = new LocalStorage({
```

Comment:
this being called `LocalStorage` and being imported from `/web` always scares me

Reply:
renamed and moved